### ISSUE TITLE
webrtc-rs-rtp was yanked

### ISSUE THREAD

**Issue by nirvdrum** [OpenXbox/xcloud-rs#3]
I just cloned the project (rev 2f4e0d498958506fef1108d96ca57b54457af5c9) and was unable to build it because _gamestreaming_'s [webrtc-rs-rtp dependency](https://crates.io/crates/webrtc-rs-rtp) cannot be found. [It was yanked](https://crates.io/crates/webrtc-rs-rtp/versions) and _gamestreaming_ does not have a Cargo lock file causing a search for the most recent version to fail.

**Comment by nirvdrum** [comment]
Looking at the org page in GitHub, I saw there was activity in this project only a day ago, but the _master_ branch hasn't been updated since 13-03-2021. That led me to discover _develop_ branch, which [swaps out the dependency](1b1ae59116ddc758914f56de10d96aff0e59a29f).

I don't know what the project's policy is with regards to when things get marked as fixed. It's fixed on the _develop_ branche, but the issue currently appears on the default branch (_master_) and that's what the README's instructions use.

**Comment by tuxuser** [comment]
Hey @nirvdrum, thanks for your interest. This project is unfortunately not in a usable state at all.
As you noticed, I just recently upgraded the dependencies and fixed up the resulting API changes.

Now the next step is to soft-deprecate the `gamestreaming` module by renaming it to `gamestreaming_native` and freeze further development for now - as this native protocol is not fully researched yet.

Instead, a new modue `gamestreaming_webrtc` will be introduced, which will base off the implementation by @unknownskl (https://github.com/unknownskl/xbox-xcloud-player) - but instead of using the internet browser WebRTC functionality, I would like to use https://github.com/webrtc-rs/webrtc

TL;DR: It's a work-in-progress codebase :)

PS: Good hint on the `Cargo.lock` file!

**Comment by nirvdrum** [comment]
Ahh, no worries. Thanks for the info. Based on the I was under the impression the openxbox.org site, I was under the impression this was further along. Although, I was also surprised because the last time I looked into this topic, I couldn't find any information. I was hoping to learn more about the protocol by digging into the code. Incidentally, I started by looking at xbox-cloud-player, but an implementation that didn't require Node or Electron was appealing.

I'd be happy to help contribute. I don't know much about WebRTC, but this could be an interesting learning opportunity.

**Comment by tuxuser** [comment]
Update on the state of the project:

- PoC works (see https://github.com/OpenXbox/xcloud-rs/blob/develop/gamestreaming_webrtc/src/bin/client-webrtc.rs - It's quite messy cause atm its thrown-together example code from webrtc-rs repo)
- Project structure was cleaned up and refactored

What currently works in the PoC:
- Creating a connection to XCloud
- Downloading the list of playable titles
- Automatically choose the first title in the (alphabetical) list
- Starting a WebRTC session
- Once session is established, write audio/video stream to video.mkv/audio.ogg respectively

Whats missing:
- Handling of the DataChannels to enable Gamepad input. The logic for that (what to send, what to receive) is already enabled in the base library.
- Rendering audio/video to a surface

With that in mind, I am targeting a second approach for a client implementation -> GStreamer (instead of webrtc-rs, see: https://docs.rs/gstreamer/latest/gstreamer/)

Why?
- GStreamer should be able to do everything that webrtc-rs can + it brings its own implementation to render video / playback audio. Also it does decoding by itself.
- Only the DataChannels would be handled then in the background

If you like to have a chat, about areas to contribute, jump to [Discord](https://openxbox.org/discord)

PS: PoC can be invoked by:

```
# Auth and get tokens.json
cargo run --bin auth-webview --features=webview
# Start PoC client
cargo run --bin client-webrtc --features="xal webrtc-rs"
```
